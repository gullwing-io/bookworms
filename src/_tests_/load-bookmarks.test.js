import {
  fetchBookmarkConfig,
  shouldFetchFromLocal,
  fetchLocaleBookmarkYAML,
  fetchRemoteBoomarkYMAL,
  returnResponseOn2xx,
  returnResponseAsObject,
} from "../load-bookmarks";

describe("loading bookmark config", () => {
  describe("fetchBookmarkConfig", () => {
    test("should return locale bookmarks as JSON", async () => {
      expect(
        await fetchBookmarkConfig("./demo/config/bookmarks.yaml")
      ).toMatchSnapshot();
    });
    test("should return remote bookmarks as JSON", async () => {
      expect(
        await fetchBookmarkConfig(
          "https://raw.githubusercontent.com/thearegee/bookworms/main/demo/config/bookmarks.yaml"
        )
      ).toMatchSnapshot();
    });
    test("should return locale bookmarks as YAML", async () => {
        expect(
          await fetchBookmarkConfig("./demo/config/bookmarks.yaml", true)
        ).toMatchSnapshot();
      });
      test("should return remote bookmarks as YAML", async () => {
        expect(
          await fetchBookmarkConfig(
            "https://raw.githubusercontent.com/thearegee/bookworms/main/demo/config/bookmarks.yaml", true
          )
        ).toMatchSnapshot();
      });
  });
  describe("shouldFetchFromLocal", () => {
    test("should return true if path is local filesystem", () => {
      expect(shouldFetchFromLocal("./demo/config/bookmarks.yaml")).toBeTruthy();
    });
    test("should return false if path is remote http", () => {
      expect(
        shouldFetchFromLocal(
          "https://raw.githubusercontent.com/thearegee/bookworms/main/demo/config/bookmarks.yaml"
        )
      ).toBeFalsy();
    });
  });
  describe("fetchLocaleBookmarkYAML", () => {
    test("should return JSON from local YMAL", () => {
      expect(
        fetchLocaleBookmarkYAML("./demo/config/bookmarks.yaml")
      ).toMatchSnapshot();
    });
  });
  describe("fetchRemoteBoomarkYMAL", () => {
    test("should return JSON from http request", async () => {
      expect(
        await fetchRemoteBoomarkYMAL(
          "https://raw.githubusercontent.com/thearegee/bookworms/main/demo/config/bookmarks.yaml"
        )
      ).toMatchSnapshot();
    });
  });
  describe("returnResponseOn2xx", () => {
    test("should throw with statusCode 100", () => {
      try {
        returnResponseOn2xx(
          "hello: world",
          100,
          "https://raw.githubusercontent.com/thearegee/bookworms/main/demo/config/bookmarks.yaml"
        );
        expect(true).toBe(false);
      } catch (error) {
        expect(error.message).toEqual(
          "https://raw.githubusercontent.com/thearegee/bookworms/main/demo/config/bookmarks.yaml returned 100"
        );
      }
    });
    test("should throw with statusCode 300", () => {
      try {
        returnResponseOn2xx(
          "hello: world",
          300,
          "https://raw.githubusercontent.com/thearegee/bookworms/main/demo/config/bookmarks.yaml"
        );
        expect(true).toBe(false);
      } catch (error) {
        expect(error.message).toEqual(
          "https://raw.githubusercontent.com/thearegee/bookworms/main/demo/config/bookmarks.yaml returned 300"
        );
      }
    });
    test("should return object with string of JSON", () => {
      try {
        const response = returnResponseOn2xx(
          "hello: world",
          200,
          "https://raw.githubusercontent.com/thearegee/bookworms/main/demo/config/bookmarks.yaml"
        );
        expect(response).toEqual({
          body: "hello: world",
          path: "https://raw.githubusercontent.com/thearegee/bookworms/main/demo/config/bookmarks.yaml",
        });
      } catch (error) {
        expect(true).toBe(false);
      }
    });
  });
  describe("returnResponseAsObject", () => {
    test("should return throw with invalid response error", () => {
      try {
        returnResponseAsObject(
          undefined,
          "https://raw.githubusercontent.com/thearegee/bookworms/main/demo/config/bookmarks.yaml"
        );
        expect(true).toBe(false);
      } catch (error) {
        expect(error.message).toEqual(
          "https://raw.githubusercontent.com/thearegee/bookworms/main/demo/config/bookmarks.yaml returned invalid bookworms bookmarks structure"
        );
      }
    });
    test("should return object with string of JSON", () => {
      try {
        const response = returnResponseAsObject(
          "{'hello':'world'}",
          "https://raw.githubusercontent.com/thearegee/bookworms/main/demo/config/bookmarks.yaml"
        );
        expect(response).toEqual({ hello: "world" });
      } catch (error) {
        expect(true).toBe(false);
      }
    });
    test("should return object with object", () => {
      try {
        const response = returnResponseAsObject(
          { hello: "world" },
          "https://raw.githubusercontent.com/thearegee/bookworms/main/demo/config/bookmarks.yaml"
        );
        expect(response).toEqual({ hello: "world" });
      } catch (error) {
        expect(true).toBe(false);
      }
    });
    test("should return object with YAML", () => {
      try {
        const response = returnResponseAsObject(
          "hello: world",
          "https://raw.githubusercontent.com/thearegee/bookworms/main/demo/config/bookmarks.yaml"
        );
        expect(response).toEqual({ hello: "world" });
      } catch (error) {
        expect(true).toBe(false);
      }
    });
  });
});
