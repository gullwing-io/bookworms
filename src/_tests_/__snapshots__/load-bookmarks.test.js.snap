// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loading bookmark config fetchBookmarkConfig should return locale bookmarks as JSON 1`] = `
Object {
  "body": Object {
    "description": "These are sample bookmarks to teach you how Bookworms works",
    "folders": Array [
      Object {
        "description": "This is to describe the folder structure",
        "folders": Array [
          Object {
            "bookmarks": Array [
              Object {
                "description": "this is used to describe the bookmark",
                "href": "https://www.mywebsite.com",
                "label": "sample url 1",
              },
            ],
            "description": "This is to describe the sub folder structure",
            "label": "sub folder 1",
          },
        ],
        "label": "folder 1",
      },
      Object {
        "folders": Array [
          Object {
            "bookmarks": Array [
              Object {
                "description": "this is used to describe the bookmark",
                "href": "https://www.mywebsite.com",
                "label": "sample url 2",
              },
              Object {
                "description": "this is used to describe the bookmark",
                "href": "https://www.mywebsite.com",
                "label": "sample url 3",
              },
            ],
            "label": "sub folder 2",
          },
          Object {
            "bookmarks": Array [
              Object {
                "description": "this is used to describe the bookmark",
                "href": "https://www.mywebsite.com",
                "label": "sample url 4",
              },
              Object {
                "description": "this is used to describe the bookmark",
                "href": "https://www.mywebsite.com",
                "label": "sample url 5",
              },
            ],
            "label": "sub folder 3",
          },
        ],
        "label": "folder 2",
      },
    ],
    "label": "Bookworms",
  },
  "type": "local",
}
`;

exports[`loading bookmark config fetchBookmarkConfig should return locale bookmarks as YAML 1`] = `
Object {
  "body": "label: Bookworms
description: These are sample bookmarks to teach you how Bookworms works
folders:
  - 
    label: folder 1
    description: This is to describe the folder structure
    folders:
      - 
        label: sub folder 1
        description: This is to describe the sub folder structure
        bookmarks:
          - 
            label: sample url 1
            description: this is used to describe the bookmark
            href: https://www.mywebsite.com
  - 
    label: folder 2
    folders:
      - 
        label: sub folder 2
        bookmarks:
          - 
            label: sample url 2
            description: this is used to describe the bookmark
            href: https://www.mywebsite.com
          - 
            label: sample url 3
            description: this is used to describe the bookmark
            href: https://www.mywebsite.com
      - 
        label: sub folder 3
        bookmarks:
          - 
            label: sample url 4
            description: this is used to describe the bookmark
            href: https://www.mywebsite.com
          - 
            label: sample url 5
            description: this is used to describe the bookmark
            href: https://www.mywebsite.com",
  "type": "local",
}
`;

exports[`loading bookmark config fetchBookmarkConfig should return remote bookmarks as JSON 1`] = `
Object {
  "body": Object {
    "description": "These are sample bookmarks to teach you how Bookworms works",
    "folders": Array [
      Object {
        "description": "This is to describe the folder structure",
        "folders": Array [
          Object {
            "bookmarks": Array [
              Object {
                "description": "this is used to describe the bookmark",
                "href": "https://www.mywebsite.com",
                "label": "sample url 1",
              },
            ],
            "description": "This is to describe the sub folder structure",
            "label": "sub folder 1",
          },
        ],
        "label": "folder 1",
      },
      Object {
        "folders": Array [
          Object {
            "bookmarks": Array [
              Object {
                "description": "this is used to describe the bookmark",
                "href": "https://www.mywebsite.com",
                "label": "sample url 2",
              },
              Object {
                "description": "this is used to describe the bookmark",
                "href": "https://www.mywebsite.com",
                "label": "sample url 3",
              },
            ],
            "label": "sub folder 2",
          },
          Object {
            "bookmarks": Array [
              Object {
                "description": "this is used to describe the bookmark",
                "href": "https://www.mywebsite.com",
                "label": "sample url 4",
              },
              Object {
                "description": "this is used to describe the bookmark",
                "href": "https://www.mywebsite.com",
                "label": "sample url 5",
              },
            ],
            "label": "sub folder 3",
          },
        ],
        "label": "folder 2",
      },
    ],
    "label": "Bookworms",
  },
  "type": "remote",
}
`;

exports[`loading bookmark config fetchBookmarkConfig should return remote bookmarks as YAML 1`] = `
Object {
  "body": "label: Bookworms
description: These are sample bookmarks to teach you how Bookworms works
folders:
  - label: folder 1
    description: This is to describe the folder structure
    folders:
      - label: sub folder 1
        description: This is to describe the sub folder structure
        bookmarks:
          - label: sample url 1
            description: this is used to describe the bookmark
            href: https://www.mywebsite.com
  - label: folder 2
    folders:
      - label: sub folder 2
        bookmarks:
          - label: sample url 2
            description: this is used to describe the bookmark
            href: https://www.mywebsite.com
          - label: sample url 3
            description: this is used to describe the bookmark
            href: https://www.mywebsite.com
      - label: sub folder 3
        bookmarks:
          - label: sample url 4
            description: this is used to describe the bookmark
            href: https://www.mywebsite.com
          - label: sample url 5
            description: this is used to describe the bookmark
            href: https://www.mywebsite.com",
  "type": "remote",
}
`;

exports[`loading bookmark config fetchLocaleBookmarkYAML should return JSON from local YMAL 1`] = `
"label: Bookworms
description: These are sample bookmarks to teach you how Bookworms works
folders:
  - 
    label: folder 1
    description: This is to describe the folder structure
    folders:
      - 
        label: sub folder 1
        description: This is to describe the sub folder structure
        bookmarks:
          - 
            label: sample url 1
            description: this is used to describe the bookmark
            href: https://www.mywebsite.com
  - 
    label: folder 2
    folders:
      - 
        label: sub folder 2
        bookmarks:
          - 
            label: sample url 2
            description: this is used to describe the bookmark
            href: https://www.mywebsite.com
          - 
            label: sample url 3
            description: this is used to describe the bookmark
            href: https://www.mywebsite.com
      - 
        label: sub folder 3
        bookmarks:
          - 
            label: sample url 4
            description: this is used to describe the bookmark
            href: https://www.mywebsite.com
          - 
            label: sample url 5
            description: this is used to describe the bookmark
            href: https://www.mywebsite.com"
`;

exports[`loading bookmark config fetchRemoteBoomarkYMAL should return JSON from http request 1`] = `
Object {
  "body": "label: Bookworms
description: These are sample bookmarks to teach you how Bookworms works
folders:
  - label: folder 1
    description: This is to describe the folder structure
    folders:
      - label: sub folder 1
        description: This is to describe the sub folder structure
        bookmarks:
          - label: sample url 1
            description: this is used to describe the bookmark
            href: https://www.mywebsite.com
  - label: folder 2
    folders:
      - label: sub folder 2
        bookmarks:
          - label: sample url 2
            description: this is used to describe the bookmark
            href: https://www.mywebsite.com
          - label: sample url 3
            description: this is used to describe the bookmark
            href: https://www.mywebsite.com
      - label: sub folder 3
        bookmarks:
          - label: sample url 4
            description: this is used to describe the bookmark
            href: https://www.mywebsite.com
          - label: sample url 5
            description: this is used to describe the bookmark
            href: https://www.mywebsite.com",
  "path": "https://raw.githubusercontent.com/thearegee/bookworms/main/demo/config/bookmarks.yaml",
}
`;
